### PR TITLE
feat: forge coverage

### LINKED ISSUE #99: dapptools-style coverage
Implemented at: https://github.com/dapphub/dapptools/blob/728a9245fa5f78589b0cedec0ade2da5433ad792/src/hevm/src/EVM/UnitTest.hs#L251-L382

Example: https://twitter.com/dapptools/status/1435973810545729536

### PR DESCRIPTION
**Pull request by onbjerg**
Adds test coverage support to Forge.

### Usage

> **You cannot currently run this locally easily**
>
> This PR depends on https://github.com/gakonst/ethers-rs/pull/1201, so to run locally you must clone `ethers-rs` from that branch
> and update the patch section in `Cargo.toml` to point to your patched ethers. Then, run `foundryup -p .`.

Simply run `forge coverage` to get a coverage summary.

An LCOV reporter is also included: `forge coverage --report lcov`. Currently the LCOV file is written to `./lcov.info` but this is going to move.

### Background

I've had to rewrite this feature many times.. I wrote what I thought would be a good structure, only to find out later that something would not work and/or fit into the current structure, so I scrapped most of it and tried again.

Initially, the thought was to analyse `JUMPI` and `JUMPDEST`s to find branches (i.e. `require`, `assert`, `if` etc.), however it would not get us all the way in terms of the information we want - we want functions, branches, executable lines and statements.

The primary mode of operation for the coverage tool will be to collect the following pieces of information:

- **Functions** (done)
- **Executable lines**: lines, excluding lines that do nothing (whitespace, comments, brackets on single lines etc.)
- **Branches**: If statements, calls to require and asserts (wip)
- **Statements**: executable statements, which differ from lines in that you can have multiple statements per executable line (sort of done)

This information will be collected from the AST of the Solidity files. Then the tests are run with a special coverage collector that marks opcodes as hit as they are executed, which we later use (along with source maps) to map back to the coverage items.

Currently, I am working off of a combination of [`solidity-coverage`](https://github.com/sc-forks/solidity-coverage/) and [this Brownie blog post](https://medium.com/coinmonks/brownie-evaluating-solidity-code-coverage-via-opcode-tracing-a7cf5a92d28c) to figure out what to include in the reports and relevant data structures.

### Report types

The above information will be output either to stdout as a simple report, or as an LCOV file that you can upload to services like CodeCov. A stretch goal would be to add a HTML report as well, but since tools already exist for converting LCOV to HTML, it isn't a deal breaker if it is not in the first version.

### Compared to other coverage tools

- DappTools's coverage tool is not super great (see comments above) as it does not support outputting to standard reporting formats like LCOV, and it does not treat `require` and `assert` as branches.
- Hardhat Coverage is closer to what we want, but it's slow and it distorts gas since it instruments the code of the contracts

### Current status

Good 

- **All functions are detected**
- **Most statements are detected**, however there are some TODOs in the source code itself with questions about how deep in the AST we want to go, specifically around subexpressions.
- **I am collecting hit data** for individual instructions

Missing

- **Executable lines are not detected**: I left this for last because (fingers crossed) this would be the easiest item type to detect
- **Not all branches are detected**: I am detecting if-statements, however, I do wonder if it works as intended (e.g. should we mark chained `elseif`'s with the same branch ID?). I also do not currently detect `assert` and `require` type branches.
- **HTML reporter**: There is no HTML reporter and I consider it a stretch goal.
- **Hit data is NOT hooked up yet** so every item shows up as uncovered. My current thinking is that I want good detection, and then I'll start hooking up the hit data, but it is possible to hook it up first.

Needs change

- **LCOV reports do not have a fixed place yet**: I need to figure out where to put report data in general: I've thought about `out/coverage` (e.g. `out/coverage/lcov.info`, `out/coverage/html/index.html`), but I'm not sure yet.
- **Dependencies and test contracts are not filtered out**: We don't want to generate coverage for dependencies or the test contracts themselves
- **The summary reporter is verbose**: Currently the full path for files is displayed, but it should work more like what @mds1 had in mind in #99 

Needs clarification

- **Subexpression clarifications**: I'm not sure how deep to go for statement coverage in terms of subexpressions. See the relevant TODOs in the source code. I am also not sure if try/catch counts as a branch, or whether `Conditional` nodes count as branches (e.g. ternary statements).
- **Probably only works for Solidity >= 0.8.0** but I'm not sure yet. There have been some breaking changes in how the AST is structured (specifically, `name` used to be `nodeType` and function names etc. were in `node.attributes.name`)
- **LCOV block IDs**: LCOV branches (specifically the `BRDA` item) have a branch ID and a block ID. I'm not sure what the block ID is supposed to mean.

### Task list

**Branches**

- [ ] Detect `assert` as a branch
- [ ] Detect `require` as a branch
- [ ] Figure out if try/catch is a branch
- [ ] Figure out if `Conditional` nodes are branches

**Lines**

- [x] Detect executable lines

**Statements**

- [ ] Figure out what subexpressions to include as statement items

**Reporters**

- [ ] Add a HTML reporter like [`solidity-coverage`](https://github.com/sc-forks/solidity-coverage/)
- [ ] Bring summary reporter more in line with @mds1's idea in #99 
- [ ] Figure out if the LCOV reporter is broken or not
       I've tried running the LCOV report through `genhtml` but it complained that the file was malformed. I'm not sure if this is to be expected, or if it is just because we don't map the relevant items to source lines and that is why it is complaining

**Misc**

- [x] Hook up hit data
- [ ] Figure out what Solidity versions are supported and what versions we *should* support
- [x] Filter out dependency contracts
- [x] Filter out test contracts
- [ ] Try to break it and fix or handle those breakages

### Other

Companion PR: https://github.com/gakonst/ethers-rs/pull/1201

Closes #99

### CODE REVIEW

File: cli/src/cmd/forge/coverage.rs
Comment (mattsse) [inline cli/src/cmd/forge/coverage.rs:138]: I feel like we should add an extension trait for this

File: cli/src/cmd/forge/coverage.rs
Comment (mattsse) [inline cli/src/cmd/forge/coverage.rs:147]: see https://github.com/gakonst/ethers-rs/pull/1253

File: cli/src/cmd/forge/coverage.rs
Comment (mattsse) [inline cli/src/cmd/forge/coverage.rs:255]: let's move all of this to a sub module

File: evm/src/coverage.rs
Comment (mattsse) [inline evm/src/coverage.rs:27]: unneccessary?

File: evm/src/executor/inspector/coverage.rs
Comment (mattsse) [inline evm/src/executor/inspector/coverage.rs:24]: does this need to be sorted?

Comment (onbjerg) [reply evm/src/executor/inspector/coverage.rs:24]: Not really, but the alternative is that we hash the small integers which I think is a bigger performance penalty?

Comment (mattsse) [reply evm/src/executor/inspector/coverage.rs:24]: simple hashmap should also do it

Comment (onbjerg) [reply evm/src/executor/inspector/coverage.rs:24]: Yeah but that's what I mean - the HashMap will hash the usizes and this can be pretty slow considering how small the usizes really are. I don't think BTreeMap hashes?

File: cli/src/cmd/forge/coverage.rs
Comment (gakonst) [inline cli/src/cmd/forge/coverage.rs:109]: FYI @mattsse

Comment (gakonst) [inline cli/src/cmd/forge/coverage.rs:109]: FYI @mattsse

**Comment by brockelmore** [comment]
Random thought, likely shouldnt be in current PR but:

It would be cool to get info on storage slot writes/checks. i.e. `funcA` writes to slots 1 and 2, any test that calls `funcA` should check slots 1 and 2. This can help users identify potentially missed state transitions in their tests

**Comment by gakonst** [comment]
I think if the optimizer is On we get discrepancies between bytecode and sourcemap?

**Comment by brockelmore** [comment]
ye we do. i think best solution is to try to turn it off and if we get a stack too deep error and the user specifies to use the optimizer, we should just compile with the optimizer on and report a warning that we had to turn on the optimizer so coverage may not be accurate. otherwise coverage is impossible (and a slightly incorrect coverage report is better than no coverage report imo)

**Comment by brockelmore** [comment]
ye we do. i think best solution is to try to turn it off and if we get a stack too deep error and the user specifies to use the optimizer, we should just compile with the optimizer on and report a warning that we had to turn on the optimizer so coverage may not be accurate. otherwise coverage is impossible to run (and a slightly incorrect coverage report is better than no coverage report imo)

**Comment by brockelmore** [comment]
ye we do. i think best solution is to try to turn it off and if we get a stack too deep error and the user specifies to use the optimizer, we should just compile with the optimizer on and report a warning that we had to turn on the optimizer so coverage may not be accurate. otherwise coverage is impossible to run (and a slightly incorrect coverage report is better than no coverage report imo)

**Comment by brockelmore** [comment]
ye we do (kind of, bytecode/sourcemap is accurate for the most part but the bytecode/sourcemap/ast conversions can get mangled iirc). i think best solution is to try to turn it off and if we get a stack too deep error and the user specifies to use the optimizer, we should just compile with the optimizer on and report a warning that we had to turn on the optimizer so coverage may not be accurate. otherwise coverage is impossible to run (and a slightly incorrect coverage report is better than no coverage report imo)

**Comment by gakonst** [comment]
I do wonder why does solc not emit an "optimized" sourcemap when the bytecode is also optimized..

**Comment by brockelmore** [comment]
> I do wonder why does solc not emit an "optimized" sourcemap when the bytecode is also optimized..

to be clear, it tries its best to do so. I think it just mangles the bytecode to a point that so much of the bytecode doesnt actually map to anything in the source. So it technically does. The primary issue i believe is that a lot of the optimizations that are done are done via "peephole" optimizations, optimizations done on the bytecode itself, which means it doesnt track to the AST so the AST is only partially optimized relative to the bytecode.

**Comment by gakonst** [comment]
👀  it's happening

**Comment by onbjerg** [comment]
Hey guys, an update.

First of all, to everyone following the issue, sorry for the radio silence and lack of activity. I've both had an attempt at something more complex (in order to address an issue I think might come up), which I've put on hold for now, and also had some family related issues that impacted my mental health.

On the latest commit before this comment I see that a lot of things in lil-web3 is covered:

image

I still need to:

- Detect executable lines
- Collect branch hit data
- Test the LCOV reporter

That seems like a shorter to do list, but the current thinking is that I get this into a useable experimental state and create issues separately for the remaining items, such as the HTML reporter, questions around some of the statement nodes, and figuring out what we support, what we should support, and what we can support.

I've noticed some things that I need to address before this can be merged:

- Libraries currently are not covered. I think this is because source maps are missing for these for some reason, so I'm trying to figure out why that is
- `receive()` functions are not marked as covered - I think this is because of some optimization Solidity does that I don't currently handle correctly
- The dependency and test contract checks are not super robust, and I probably need to filter on an AST level for test contracts instead of on a file level: projects like Solmate usually create concrete implementations of their abstract contracts in their test files in order to test functionality - this is not currently handled

**Comment by onbjerg** [comment]
Update again, I'm currently working on hooking up branch hit data and concurrently checking out issue #2 I wrote about (the `receive()` call). I think the `receive()` function is optimized away in some situations since it essentially just disables the value transfer check, and in those cases there is no source mapping. I think it might be fine to leave that case out for now, but I want to make sure that it works as expected for `receive()` functions that have a function body

**Comment by gakonst** [comment]
I think we need a rebase per the recent refactor by @mattsse. Should we make `0/0`'s display as not-green so that it's not confusing?

> That seems like a shorter to do list, but the current thinking is that I get this into a useable experimental state and create issues separately for the remaining items, such as the HTML reporter, questions around some of the statement nodes, and figuring out what we support, what we should support, and what we can support

I'm supportive of this, esp. the HTML reporter and edge cases around various nodes. e.g. in the Receive case, if you think it's a lot of work, let's track separately and can document it as a known limitation

**Comment by onbjerg** [comment]
image

Getting there - some items are considered uncovered even though they are. This is related to how I pick anchors (see 3bb07ce and 5742f53), since I am only considering the runtime source map

**Comment by gakonst** [comment]
Everytime I see a commit w progress on this PR I jump out of my seat lol

Do you have a 1-paragraph explanation of "anchors" and why they help here?

**Comment by onbjerg** [comment]
@gakonst Each AST node has an attached source range: a start byte and (optionally) a length. Previously I would take each opcode we executed, look it up in the source map and see if the offset described in the source map matched the start byte of one of our coverage items.

This doesn't always work: sometimes the opcode that should mark an item as covered might be somewhere else in the range, and sometimes an opcode actually marks multiple items as covered because of the optimizer. Anchors are essentially opcodes that we pick that would mark items as covered.

Another benefit of the anchor approach is that we can change what anchor we pick for more complicated items such as branches. For example we might want to pick a specific type of opcode within the source range (like a JMPI or something).

Hope what I wrote made sense, it's a little hard to explain lol

**Comment by gakonst** [comment]
This makes sense (also copying @jpopesculian and @rkrasiuk who seem to understand the AST stuff better than I frmo the work on forge fmt in case they have thoughts / comments).


Separately:

> It considers if statements with no else block to be fully covered even if the "false" branch is never hit (we need to create a "virtual" one)

This makes sense.

> It doesn't consider require and assert... These are more complicated

What makes them more complicated? Doesn't each require/assert(bool) translate to an `if (bool)` statement?

**Comment by onbjerg** [comment]
>What makes them more complicated? Doesn't each require/assert(bool) translate to an if (bool) statement and is like a branch?

Ideally for branches we want to have three states: not executed, one path executed (either true or false) or both branches executed (i.e. covered).

Currently for if statements there are two states: not executed or true branch executed which is not ideal. I think we probably want to signal to the developer if they only covered 1 possible path in their test.

For require and assert though, all of this is marked by specific opcodes, and I need to figure out which. The general layout seems to be:

```
<some assertion>
PUSH <pc at which we continue execution if assertion holds>
JUMPI
<setup revert and revert>
```

So my assumption is that the false branch would be right after JUMPI. For the true branch, I'm not sure - maybe I need to record what value is on the stack every time we encounter a `JUMPI` so I can look it up later and check if we both got a 0 and a non-0 value? :/

**Comment by gakonst** [comment]
(calling in the cavalry) @MrChico sorry to bother - do you have any thoughts on how we can handle require/asserts in a nice way?

**Comment by onbjerg** [comment]
Also of note - this is just for 0.8.13. Not sure if this is remarkably different for other versions, but we do run the risk of this being fragile

**Comment by onbjerg** [comment]
Moving out of draft - still a lot to do, will create separate issues to track.

**Comment by gakonst** [comment]
> 
Also of note - this is just for 0.8.13. Not sure if this is remarkably different for other versions, but we do run the risk of this being fragile

Yep so worth testing with latest, and then see how it works in older versions and document limitations if any. We don't need to worry TOO much about backwards compat imo, but is a nice to have.

File: cli/src/cmd/forge/coverage.rs
Comment (mattsse) [inline cli/src/cmd/forge/coverage.rs:253]: why execute this on a new thread and immediately join it again?

File: evm/src/coverage/mod.rs
Comment (mattsse) [inline evm/src/coverage/mod.rs:31]: not needed?

File: cli/src/cmd/forge/coverage.rs
Comment (onbjerg) [reply cli/src/cmd/forge/coverage.rs:253]: This is how forge test works too - after some changes where we stream in results of test runs we have to run it in a new thread, otherwise we get locked I think?

File: cli/src/cmd/forge/coverage.rs
Comment (mattsse) [inline cli/src/cmd/forge/coverage.rs:77]: nit: we now have silent in `CorebuildArgs` so this should print via the `p_println!` macro

File: cli/src/cmd/forge/coverage.rs
Comment (mattsse) [inline cli/src/cmd/forge/coverage.rs:152]: there's `FoundryPathExt` trait that has `FoundryPathExt::is_sol_test`

File: cli/src/cmd/forge/coverage.rs
Comment (mattsse) [inline cli/src/cmd/forge/coverage.rs:175]: let's wrap all io errors with their paths, because the path is not included in `io::Error`

Comment (mattsse) [inline cli/src/cmd/forge/coverage.rs:175]:
```suggestion
                        std::fs::read_to_string(&path)?,
```

File: cli/src/cmd/forge/coverage.rs
Comment (mattsse) [inline cli/src/cmd/forge/coverage.rs:259]: should these all be simple functions, something like:

```rust
impl SummaryReporter {
 pub fn create(map){}
}
```

File: evm/src/coverage/mod.rs
Comment (mattsse) [inline evm/src/coverage/mod.rs:38]: nit: `impl Into<PathBuf>`

File: evm/src/coverage/mod.rs
Comment (mattsse) [inline evm/src/coverage/mod.rs:59]: move to fn docs

File: evm/src/coverage/mod.rs
Comment (mattsse) [inline evm/src/coverage/mod.rs:129]: nice

File: evm/src/coverage/mod.rs
Comment (mattsse) [inline evm/src/coverage/mod.rs:331]: nit: move this to `AddAssign` trait impl 

```rust
impl AddAssign/*perhaps <&CoverageSummary>*/ for CoverageSummary { }
```

File: evm/src/coverage/visitor.rs
Comment (mattsse) [inline evm/src/coverage/visitor.rs:10]: this is a Visitor for the `AST` object that solc includes in the output, right?

File: evm/src/coverage/visitor.rs
Comment (mattsse) [inline evm/src/coverage/visitor.rs:50]: should we make this more ergonomic on the `Node`, with a `enum Attribute` and `enum Kind` or something?

File: evm/src/executor/builder.rs
Comment (mattsse) [inline evm/src/executor/builder.rs:160]: considering how this will be configured, perhaps we should make this `set_coverage(mut self, bool)` so we can easily chain this to the `Builder`

File: forge/src/coverage.rs
Comment (mattsse) [inline forge/src/coverage.rs:31]: redundant

File: forge/src/coverage.rs
Comment (mattsse) [inline forge/src/coverage.rs:9]: looking at the `match self.report` earlier

we don't really make use of that, right?

but nice to have

File: forge/src/coverage.rs
Comment (mattsse) [inline forge/src/coverage.rs:17]: the list of reporters may grow and become more complex, so I think we should consider moving all of them to their own module

File: forge/src/coverage.rs
Comment (mattsse) [inline forge/src/coverage.rs:80]: consider `destination: &'a mut (dyn Write + 'a)` this is what the `std::Formatter` does.

the benefit of this is, that you don't need to hand over ownership to the reporter and keep using the destination when the reporter is done

File: forge/src/multi_runner.rs
Comment (mattsse) [inline forge/src/multi_runner.rs:177]: same as above

File: forge/src/multi_runner.rs
Comment (mattsse) [inline forge/src/multi_runner.rs:297]: if we change the builder function we can get rid of this and do

```rust
.set_coverage(self.coverage)
....

```

**Review by mattsse — CHANGES_REQUESTED** [review]

**Comment by tuoomz** [comment]
Any idea when this will be merged

File: evm/src/coverage/visitor.rs
Comment (onbjerg) [reply evm/src/coverage/visitor.rs:10]: yes

File: evm/src/coverage/visitor.rs
Comment (onbjerg) [reply evm/src/coverage/visitor.rs:50]: I considered it initially when doing my first AST PR but it was hard to figure out a nice way to do it within the constraints of how it was already designed. I couldn't really see a good way to add this extra data without having a big struct where half of the fields were empty for most nodes without creating node-specific types and moving more towards something like what solang is doing (e.g. a "root" AST node with children that are `Box<something>` and so on all the way down) which might have been to expensive? Interested to hear your thoughts - the visitor here is pretty simple so adjusting it shouldn't be too much work.

File: evm/src/coverage/mod.rs
Comment (mattsse) [inline evm/src/coverage/mod.rs:26]: General question: Is there an advantage to having this sorted compared to HashMap

File: evm/src/coverage/visitor.rs
Comment (mattsse) [reply evm/src/coverage/visitor.rs:50]: I see, 
sounds reasonable, we could however think of adding an enum (Known)Attribute that's `AsRef<str>` so we can do `node.attribute(Attribute::Abstract)`

but no real benefit to that I'd assume

File: evm/src/coverage/visitor.rs
Comment (mattsse) [inline evm/src/coverage/visitor.rs:369]: there's `str::lines()` which should do this, except platform (windows) agnostic

**Review by mattsse — APPROVED** [review]
smol nit

lgtm!

how can we test this properly?

**Comment by onbjerg** [comment]
> smol nit
> 
> lgtm!
> 
> how can we test this properly?

Not sure how to best do automated testing for it yet, my best guess is to do some sort of snapshot testing against external repos - so, we run `forge coverage` for external repos, snapshot the output and if it differs we'd have to manually check that it looks ok. For some parts of the code though we can probably be a bit smarter about it (e.g. the format of the reporters, parts of the visitor).

There's still some way to go before this is dependable (see above) but those will be in other PRs

File: evm/src/coverage/mod.rs
Comment (onbjerg) [reply evm/src/coverage/mod.rs:26]: Mmm, probably not, no

**Comment by mattsse** [comment]
SGTM

**Pull request merged**
